Support for N:M nets

mundya opened the pull request:
Corrects `nengo_spinnaker` so that it no longer (incorrectly) relies on Rig providing N:M nets.
- `nengo_spinnaker.netlist.Net` now explicitly represents N:M nets.
  - The builder is modified to produce N:M nets.
- `nengo_spinnaker.netlist.utils` provides methods to generate Rig for placement and routing. Nets for routing take account of placement and allow (N-on-one-chip):M nets to be routed as one 1:net.
  - Changes to the place and route flow now means that it is not possible to trivially extract a netlist for playing with (@mossblaser), hence the methods under `utils/place_and_route.py` are removed.
- Cluster identification is greatly simplified and moved into Netlist utility methods.
  - `partition_and_cluster` is renamed to `partition`.

(@mossblaser, would you mind having a glance at this? Sorry to bother you!)

mundya committed:
Improve netlist constraints support

mundya committed:
Support for N:M nets

Corrects `nengo_spinnaker` so that it no longer (incorrectly) relies on
Rig providing N:M nets.

 - `nengo_spinnaker.netlist.Net` now explicitly represents N:M nets.
   - The builder is modified to produce N:M nets.
 - `nengo_spinnaker.netlist.utils` provides methods to generate Rig Nets
   for placement and routing. Nets for routing take account of the
   placement and allow (N-on-one-chip):M nets to be routed as one 1:M
   net.
   - Changes to the place and route flow now means that it is not
     possible to trivially extract a netlist for playing with (sorry
     @mossblaser), hence the methods under `utils/place_and_route.py`
     are removed.
 - Cluster identification is greatly simplified and moved into the
   Netlist utility methods.
   - `partition_and_cluster` is renamed to `partition`.

mossblaser commented on nengo_spinnaker/netlist/utils.py:
```diff
@@ -0,0 +1,169 @@
+        for placement in start_placements:
+            # Create a new source vertex and place it at the given placement
+            vertex = object()
+            extended_placements[vertex] = placement
```
I realise that current implementations (and probably future ones too) of the router do not actually look at `vertices_resources` and just placements for a list of vertices but strictly speaking letting these structures go out of sync is a bug.

I'd suggest one of:
1. A big scary comment and continue like you are
2. Update vertices_resources and allocations too
3. Use an arbitrarily selected existing vertex on the chip in question.

Up to you :).

mundya replied:
For option (2) would you recommend that the new vertex is assigned no cores or all the cores of the vertices that it's pretending to be... is it enough that it's in the dictionary at all?

mossblaser replied:
Strictly you should assign it no resources (since the resources you're referring to have been assigned to the original vertices). Obviously this "hints" at the wrong thing but it is at least guaranteed to be safe.

Truly exposing accurate information to the router is probably not possible in the general case for N:M nets. For instance, you could replace all the source vertices with your new artificial vertex whose resource consumption was the sum of the vertices it replaced. This immediately fails since the resource allocations could be discontinuous. It also fails if you have overlapping subsets of your vertices forming sources in N:M nets since you can't represent that information in Rig's 1:M net model. Clearly this is a deficiency (I believe one highlighted somewhere in the rambling Rig github issue) though one which presently doesn't cause any pain since the existing router doesn't use this information.

mundya replied:
How does 4593e1a strike you?

mossblaser replied:
Looks legit!

mundya replied:
Thanks!

mossblaser commented on nengo_spinnaker/netlist/utils.py:
```diff
@@ -0,0 +1,169 @@
+                cluster_ids = set(vx.cluster for vx in original_net.sources
+                                  if placements[vx] == placement)
+                assert len(cluster_ids) == 1, "Inconsistent cluster IDs"
+                cluster_id = next(iter(cluster_ids)) or 0  # Get the ID
```
Given the assertion above, the "or 0" shouldn't be necessary.

mundya replied:
It is possible for `cluster_ids` to equal `{None}`; the `or 0` catches this case.  If `cluster_ids` were empty then `next(iter(cluster_ids))` would raise `StopIteration` and the `or` wouldn't save us.

mossblaser commented on nengo_spinnaker/netlist/objects.py:
```diff
@@ -0,0 +1,71 @@
+class Net(object):
```
Since this is no longer a subclass nor trivial extension of Rig's "net" concept I'd prefer if this had an alternative name to distinguish it.

mossblaser commented on tests/builder/test_builder.py:
```diff
@@ -516,7 +516,7 @@ def test_single_vertices(self):
         assert len(netlist.nets) == 1
         for net in netlist.nets:
-            assert net.source is vertex_a
+            assert net.sources[0] is vertex_a
```
Assert the length too?

mossblaser commented on tests/builder/test_builder.py:
```diff
@@ -613,7 +613,7 @@ def test_multiple_sink_vertices(self):
         assert len(netlist.nets) == 1
         for net in netlist.nets:
-            assert net.source is vertex_a
+            assert net.sources[0] is vertex_a
```
(Again) Assert the length too?

mossblaser commented on tests/netlist/test_netlist_netlist.py:
```diff
@@ -0,0 +1,224 @@
+
+
+def test_load_application():
+    """Test the steps involved in loading an application to a SpiNNaker
```
"loading an application **onto** a SpiNNaker Machine"?

mossblaser commented on tests/netlist/test_netlist_utils.py:
```diff
@@ -0,0 +1,225 @@
+    # Create the vertices
+    a = [object(), object()]
+    b = [object(), object()]
+    c = object()
```
This naming scheme is slightly confusing. If you can be bothered doing

```
a = object()
b = object()
c = object()
d = object()
e = object()

ab_cd = Net([a, b], [c, d], 1.0, None)
c_e = Net(c, e, 2.0, None)
```

etc. then it would (in my eyes!) make this test much more obvious!

mossblaser commented on tests/netlist/test_netlist_utils.py:
```diff
@@ -0,0 +1,225 @@
+            assert net.source is b[0]
+            assert net.sinks == [c]
+            assert net.weight == bc.weight
+
```
Also assert `seen_sources` is complete afterwards?... I guess it is kind-of redundant but it gets the point across.

mossblaser commented on tests/netlist/test_netlist_utils.py:
```diff
@@ -0,0 +1,225 @@
+
+    # Create some placements:
+    #  - A[0, 1] placed on the same chip
+    #  - B[0, 1] placed on different chips
```
A -> a, B -> b

mossblaser commented:
Aside from the one point about updating the P&R data structures consistently, the rest is more or less just nit-picking and this LGTM.

mundya committed:
Improvements due to @mossblaser

 - Renames `nengo_spinnaker.netlist.Net` to `NMNet` to indicate that it
   does more than the standard Rig Net.
 - Clarify many of the tests

mundya committed:
Keep Rig arguments in sync

As strongly recommended by @mossblaser

This ensures that `vertices_resources` and `allocations` are updated
alongside `placements` when generating pseudo-vertices to act as the
start of N:M nets.

mundya commented:
@mossblaser, are you happy that this doesn't do anything unexpected with Rig and looks sane?

If so, I'll squash and merge.

Thank you :)

mundya committed:
Merge branch 'master' into multisource-nets

mundya committed:
Merge branch 'master' into multisource-nets

mossblaser commented:
Looks sane to me!

I think N:M nets are going to have to become a thing...

mundya commented:
> I think N:M nets are going to have to become a thing...

Yeah...

mundya commented:
Thanks :)
